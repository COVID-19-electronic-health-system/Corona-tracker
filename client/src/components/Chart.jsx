import React, { useEffect, useRef, useCallback, useState } from 'react';
import { useSelector, connect } from 'react-redux';
import * as chartJs from 'chart.js';
import PropTypes from 'prop-types';

const Chart = props => {
  const { chartType, observations } = props;
  const temperatureData = useSelector(state => state.temperatureReducer.temperature);
  const [chartConfig, setChartConfig] = useState({
    type: chartType,
    data: {
      labels: temperatureData.labels,
      datasets: [
        {
          label: 'Patient temperature',
          backgroundColor: '#4760ff',
          data: temperatureData.values,
        },
      ],
    },
    options: {
      scales: {
        yAxes: [
          {
            ticks: {
              max: 110,
              min: 80,
            },
          },
        ],
      },
    },
  });

  const chartContainer = useRef(null);

  const updateChartConfig = useCallback(() => {
    const map = new Map();
    observations.map(observation => {
      const date = new Date(observation.date).toISOString().slice(0, 10);

      let temp;
      if (
        !observation.physical.feverSeverity ||
        observation.physical.feverSeverity === '' ||
        !parseInt(observation.physical.feverSeverity, 10) > 0
      ) {
        temp = 0;
      } else {
        temp = parseInt(observation.physical.feverSeverity, 10);
      }

      if (map.has(date)) {
        if (map.get(date).temp !== 0) {
          map.get(date).temp = (map.get(date).temp + temp) / (map.get(date).numObservations + 1);
        }
        map.set(temp, {
          temp,
          date,
        });
      } else {
        map.set(date, {
          temp,
          date,
        });
      }
      return true;
    });

    const dates = [];
    const temps = [];
    map.forEach(entry => {
      dates.push(entry.date);
      temps.push(entry.temp);
    });

    setChartConfig(config => {
      const newConfig = config;
      newConfig.data.labels = dates;
      newConfig.data.datasets[0].data = temps;
      return newConfig;
    });
  }, [observations]);

  useEffect(() => {
    updateChartConfig();
    // eslint-disable-next-line no-unused-vars
    const chart = new chartJs.Chart(chartContainer.current, chartConfig);
  }, [chartContainer, chartConfig, updateChartConfig]);

  return (
    <div>
      <canvas id="chart" ref={chartContainer} />
    </div>
  );
};

Chart.propTypes = {
  chartType: PropTypes.string.isRequired,
  observations: PropTypes.arrayOf(Object).isRequired,
};

const mapStateToProps = state => {
  return {
    observations: state.observationsReducer.observations,
  };
};

export default connect(mapStateToProps)(Chart);
