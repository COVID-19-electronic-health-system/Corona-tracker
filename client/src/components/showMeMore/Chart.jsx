import React, { useState } from 'react';
import { ResponsiveContainer, LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend } from 'recharts';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

const Chart = props => {
  const { observations, tempUnit } = props;

  const temps = observations.map(observation => {
    const datetime = new Date(observation.date).toISOString().slice(0, 10);
    let temp;
    if (
      !observation.physical.feverSeverity ||
      observation.physical.feverSeverity === '' ||
      !parseFloat(observation.physical.feverSeverity, 10) > 0
    ) {
      temp = 0;
    } else {
      temp = parseFloat(observation.physical.feverSeverity, 10);
    }

    return { date: datetime, temperature: temp };
  });

  const [temperatureData] = useState(temps);

  return (
    <ResponsiveContainer width="100%" aspect={4.0 / 1.5}>
      <LineChart
        data={temperatureData}
        margin={{
          top: 10,
          right: 50,
          left: 20,
          bottom: 5,
        }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="date" />
        <YAxis type="number" domain={tempUnit === 'fahrenheit' ? [94, 105] : [32, 42]} />
        <Tooltip />
        <Legend />
        <Line type="monotone" dataKey="temperature" stroke="#8884d8" activeDot={{ r: 8 }} />
        <Tooltip />
      </LineChart>
    </ResponsiveContainer>
  );
};

Chart.propTypes = {
  observations: PropTypes.arrayOf(PropTypes.instanceOf(Object)).isRequired,
  tempUnit: PropTypes.string.isRequired,
};

const mapStateToProps = state => {
  return {
    observations: state.observationsReducer.observations,
    tempUnit: state.onboardingReducer.tempUnit,
  };
};

export default connect(mapStateToProps)(Chart);
