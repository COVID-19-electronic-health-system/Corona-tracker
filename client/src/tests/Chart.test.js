import React from 'react';
import { render } from '@testing-library/react';

import { shallow, mount } from "enzyme";
import { Provider, useSelector } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk';

import Chart from '../components/Chart';

const mockStore = configureMockStore([thunk]);

const mockChartJsState = {dataSample: {
  labels: ["2020-02-28", "2020-03-2", "2020-03-4","2020-03-5","2020-03-9"],
  values: [80, 70, 56, 90, 55]}
}

describe('ChartJs', () => {
  it('should render a ChartJs component ', () => {
    const store = mockStore({root: {dataSample: mockChartJsState.dataSample}});
    window.HTMLCanvasElement.prototype.getContext = () => {}

    const wrapper = mount(
      <Provider store={store}>
        {/* <Chart /> */}
      </Provider>
    )
    expect(wrapper.find('Chart')).toMatchSnapshot()
  })
})