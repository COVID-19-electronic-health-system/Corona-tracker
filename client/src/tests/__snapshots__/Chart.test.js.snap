// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart should render a Chart component  1`] = `ReactWrapper {}`;

exports[`Chart should render a Chart component  2`] = `ReactWrapper {}`;

exports[`ChartJs should render a ChartJs component  1`] = `ReactWrapper {}`;
